### PR TITLE
Improve JSS performance and accuracy

### LINKED ISSUE #751: [Task] Optimization for lagging effects
## Metadata

Field | Value
--- | ---
Expected Delivery Date | N/A

## Content

It has been clear that, in particular cases, the software may be lagging, showing apparent latency. This has a significant negative effect on user experience.

This issue is expected to track the investigation on such cases. The solutions for these problems remain to be determined, since we are still searching the problems.

### LINKED ISSUE #788: [Bug] DecryptedPost.tsx trigger unnecessary component updates
# Bug Report

### System

- OS:
- OS Version:

### Browser

- Browser:
- Browser Version:

### Maskbook

- Maskbook Version:
- Installation: /* May be "Store", "ZIP", or "Self-Complied" */
- Build Commit: Optionally attach a Commit ID, if it is from an pre-release branch head

### Expected Behavior

/* Write the expected behavior here. */

### Actual Behavior

/* Write the actual behavior here. */

### How To Reproduce

/* Specify how it may be produced here. */

### PR DESCRIPTION
**Pull request by SunriseFox**
This mainly try to improve performance caused by `adoptedStyleSheets`

Known issues: 
+ For `Menu`: dialog menu is no longer a victim as it is from the same shadowroot and will be transported to the same portal. But a normal menu may still lose styles.
+ Low impact on performance: DimensionDev/Holoflows-Kit#105

close #788 
close #751

### CODE REVIEW

File: src/utils/jss/ShadowRootDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import { PortalShadowRoot } from '../../utils/jss/ShadowRootPortal'
+import { useSheetsRegistryStyles } from './renderInShadowRoot'
+
+const ResponsiveDialog = withMobileDialog({ breakpoint: 'xs' })(Dialog)
+
+export default function ShadowRootDialog(_props: any) {
+    const ref = React.useRef<HTMLDivElement>(null)
+    const styles = useSheetsRegistryStyles(ref)
+
+    // ? I need the render tree to get the shadowroot. Is the extra div must be rendered?
```
Comment (Jack-Works) [inline src/utils/jss/ShadowRootDialog.tsx]: you can set ref on the `<style>` i guess

File: src/utils/jss/renderInShadowRoot.tsx [hunk 2]
```diff
@@ -1,55 +1,119 @@
...
+export const jssRegistryMap: WeakMap<ShadowRoot, InformativeSheetsRegistry> = new WeakMap()
+
+export const useSheetsRegistryStyles = (ref: React.RefObject<Node>) => {
+    const registry = React.useCallback(() => {
+        if (!ref.current) return null
+        let current: Node = ref.current
+        // TODO: remember this lookup
+        // ! lookup the styled shadowroot
+        // ! document-fragment is node but not element
+        while (current.parentElement === current.parentNode) {
```
Comment (Jack-Works) [inline src/utils/jss/renderInShadowRoot.tsx]: https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType

File: src/utils/jss/renderInShadowRoot.tsx [hunk 3]
```diff
@@ -1,55 +1,119 @@
...
+            if (!current.parentNode) return null
+            current = current.parentNode
+        }
+        const shadowroot = current.parentNode as ShadowRoot
+        return shadowroot === portalShadowRoot ? null : jssRegistryMap.get(shadowroot)
+    }, [ref])
+    const [styleString, setStyleString] = React.useState(() => registry()?.toString())
+    React.useEffect(() => registry()?.addListener(() => setStyleString(registry()?.toString())))
```
Comment (Jack-Works) [inline src/utils/jss/renderInShadowRoot.tsx]: https://github.com/facebook/react/tree/master/packages/use-subscription

File: src/utils/jss/renderInShadowRoot.tsx [hunk 4]
```diff
@@ -1,55 +1,119 @@
...
+
+    // ? how to pass props to class component?
+    // @ts-ignore
+    ReactDOM.render(<RenderInShadowRootWrapper children={node} shadow={shadow} />, shadow as any)
+    return () => ReactDOM.unmountComponentAtNode(shadow as any)
+}
+
+// ! let jss tell us if it has made an update
+class InformativeSheetsRegistry extends SheetsRegistry {
+    private callback: Set<() => void>
```
Comment (Jack-Works) [inline src/utils/jss/renderInShadowRoot.tsx]: `private callback = new Set<(() => void)>`
Then you can get rid of the whole constructor block

File: src/utils/jss/ShadowRootPortal.tsx [hunk 5]
```diff
@@ -16,65 +14,82 @@ globalThis.getComputedStyle = new Proxy(
...
 )
 Object.assign(globalThis.window || {}, { getComputedStyle })
 
-let proxy: HTMLElement | undefined
+let proxy: ShadowRoot | undefined
+const handler: ProxyHandler<ShadowRoot> = {
+    // ! (1) to make it more like a Document
+    // ! (2) to make it more like an Element
+    get(target, property, handler) {
+        const ours = ['namespaceURI']
```
Comment (Jack-Works) [inline src/utils/jss/ShadowRootPortal.tsx]: Why list this attr here? It's not supported by Chrome now https://developer.mozilla.org/en-US/docs/Web/API/Node/namespaceURI

Comment (SunriseFox) [reply src/utils/jss/ShadowRootPortal.tsx]: react or something tried to read this so I put it here...

File: src/utils/jss/renderInShadowRoot.tsx [hunk 6]
```diff
@@ -1,55 +1,122 @@
...
+        }
+    }, [_current])
+    return useSubscription(subscription)
+}
+
 export class RenderInShadowRootWrapper extends React.PureComponent {
+    jss: Jss
+    proxy: HTMLElement
+    registry: InformativeSheetsRegistry
+    constructor(props: any) {
```
Comment (guanbinrui) [inline src/utils/jss/renderInShadowRoot.tsx]: Add props' type?

File: src/utils/jss/renderInShadowRoot.tsx [hunk 7]
```diff
@@ -1,55 +1,122 @@
...
+        // TODO: batch update
+        // ? aggregating multiple inform request to one callback is possible
+        for (const cb of this.callback) cb()
+    }
+    addListener(cb: () => void) {
+        this.callback.add(cb)
+        return () => void this.callback.delete(cb)
+    }
+    add(...args: Parameters<SheetsRegistry['add']>) {
+        super.add(...args)
```
Comment (guanbinrui) [inline src/utils/jss/renderInShadowRoot.tsx]: Looks many empty `<style />` elements generated in shadow root. Can we filter them out here?

Comment (SunriseFox) [reply src/utils/jss/renderInShadowRoot.tsx]: no. they just look empty but in fact not. you can access them by their `.sheet` property as long as it is inside a document or document-fragment (That is why style element cannot be cloned).

File: src/utils/jss/renderInShadowRoot.tsx (hunk 6 above)
Comment (SunriseFox) [reply src/utils/jss/renderInShadowRoot.tsx]: I don't quite know about typing them... Help welcome.

File: src/utils/jss/ShadowRootPortal.tsx [hunk 8]
```diff
@@ -16,65 +14,81 @@ globalThis.getComputedStyle = new Proxy(
...
 Object.assign(globalThis.window || {}, { getComputedStyle })
 
-let proxy: HTMLElement | undefined
+let proxy: ShadowRoot | undefined
+const handler: ProxyHandler<ShadowRoot> = {
+    // ! (1) to make it more like a Document
+    // ! (2) to make it more like an Element
+    get(target, property, handler) {
+        // ! (1) make react move event listeners to shadowroot instead of document
+        if (property === 'ownerDocument') return handler
```
Comment (Jack-Works) [inline src/utils/jss/ShadowRootPortal.tsx]: nice hack

File: src/utils/jss/renderInShadowRoot.tsx [hunk 9]
```diff
@@ -1,55 +1,128 @@
...
     if (shadow.mode === 'open') console.warn('Do not render with open ShadowRoot!')
-    ReactDOM.render(<RenderInShadowRootWrapper children={node} />, shadow as any)
-    livingShadowRoots.add(shadow)
-    applyAdoptedStyleSheets()
-    return () => {
-        shadow.adoptedStyleSheets = []
-        ReactDOM.unmountComponentAtNode(shadow as any)
-        livingShadowRoots.delete(shadow)
+
+    // ? how to pass props to class component?
```
Comment (Jack-Works) [inline src/utils/jss/renderInShadowRoot.tsx]: what do you mean by this question?

Comment (SunriseFox) [reply src/utils/jss/renderInShadowRoot.tsx]: @guanbinrui  resolved that. the @ts-ignore may be removed.

File: src/utils/jss/ShadowRootPortal.tsx [hunk 10]
```diff
@@ -16,65 +14,81 @@ globalThis.getComputedStyle = new Proxy(
...
-            const path = e.composedPath()
-            // @ts-ignore
-            const e2 = new e.constructor(e.type, e)
-            hackingEvents.set(e2, path)
-            shadowRoot.dispatchEvent(e2)
-            e.stopPropagation()
-            e.stopImmediatePropagation()
-        })
-    }
+// export function hackEvent(eventName: string, shadowRoot: ShadowRoot | Element) {
```
Comment (guanbinrui) [inline src/utils/jss/ShadowRootPortal.tsx]: Remove it?

File: src/utils/jss/ShadowRootPortal.tsx [hunk 11]
```diff
@@ -16,65 +14,81 @@ globalThis.getComputedStyle = new Proxy(
...
-    })
-    const nativeTarget = Object.getOwnPropertyDescriptor(Event.prototype, 'target')!.get!
-    Object.defineProperty(Event.prototype, 'target', {
-        get() {
-            if (hackingEvents.has(this)) return hackingEvents.get(this)![0]
-            return nativeTarget.call(this)
-        },
-    })
-}
+// // ? If react listen to some event, we also hack it.
```
Comment (guanbinrui) [inline src/utils/jss/ShadowRootPortal.tsx]: Remove it?

File: src/utils/jss/ShadowRootPortal.tsx (hunk 10 above)
Comment (SunriseFox) [reply src/utils/jss/ShadowRootPortal.tsx]: We don't need to clean our codebase... Leaving something for future reference is better.

Comment (Jack-Works) [reply src/utils/jss/ShadowRootPortal.tsx]: We have git history for reference

**Pull request merged**